exp/clusterresourceset/controllers/clusterresourceset_controller.go

```diff
@@ -0,0 +1,357 @@
+	Log    logr.Logger
+
+	config   *rest.Config
+	scheme   *runtime.Scheme
+	recorder record.EventRecorder
+}
+
+// +kubebuilder:rbac:groups=core,resources=events,verbs=get;list;watch;create;patch
+// +kubebuilder:rbac:groups=core,resources=secrets,verbs=get;list;watch
+// +kubebuilder:rbac:groups=core,resources=configmaps,verbs=get;list;watch
+// +kubebuilder:rbac:groups=core,resources=nodes,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=clusterresourceset.cluster.x-k8s.io,resources=*,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=clusterresourceset.cluster.x-k8s.io,resources=clusterresourcesets/status,verbs=get;update;patch
+
+func (r *ClusterResourceSetReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, reterr error) {
+	ctx := context.Background()
+
+	// Fetch the ClusterResourceSet instance.
+	clusterResourceSet := &clusterresourcesetv1.ClusterResourceSet{}
+	if err := r.Client.Get(ctx, req.NamespacedName, clusterResourceSet); err != nil {
+		if apierrors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	// Initialize the patch helper.
+	patchHelper, err := patch.NewHelper(clusterResourceSet, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	defer func() {
+		// Always attempt to Patch the ClusterResourceSet object and status after each reconciliation.
+		if err := patchHelper.Patch(ctx, clusterResourceSet); err != nil {
+			reterr = kerrors.NewAggregate([]error{reterr, err})
+		}
+	}()
+
+	// Handle normal reconciliation loop.
+	return r.reconcile(ctx, clusterResourceSet)
+}
+
+// reconcile handles cluster reconciliation.
+func (r *ClusterResourceSetReconciler) reconcile(ctx context.Context, clusterResourceSet *clusterresourcesetv1.ClusterResourceSet) (ctrl.Result, error) {
+	logger := r.Log.WithValues("ClusterResourceSet", clusterResourceSet.Name, "namespace", clusterResourceSet.Namespace)
+
+	cls, err := r.getClustersFromClusterResourceSet(ctx, clusterResourceSet)
+	if err != nil {
+		logger.Error(err, "Failed fetching clusters that matches ClusterResourceSet labels", "ClusterResourceSet", clusterResourceSet.Name)
+		return ctrl.Result{}, err
+	}
+
+	for _, cl := range cls {
+		if err := r.ApplyClusterResourceSet(ctx, cl, clusterResourceSet); err != nil {
+			logger.Error(err, "Failed applying resources to cluster", "Cluster", cl.Name)
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// getClustersFromClusterResourceSet fetches Clusters matched by the ClusterResourceSet's label selector that are in the same namespace as the ClusterResourceSet object.
+func (r *ClusterResourceSetReconciler) getClustersFromClusterResourceSet(ctx context.Context, clusterResourceSet *clusterresourcesetv1.ClusterResourceSet) ([]*clusterv1.Cluster, error) {
```

Comment:
This is a nit, so feel free to ignore, but I feel kind of redoundata to have the ClusterResourceSet for all methods of the ClusterResourceSetReconciler objects
```suggestion
func (r *ClusterResourceSetReconciler) getClusters(ctx context.Context, clusterResourceSet *clusterresourcesetv1.ClusterResourceSet) ([]*clusterv1.Cluster, error) {
```